File: by-kotlin/kt-boj/solutions/4949-균형잡힌 세상.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    while (true) {
        val text = readLine().toCharArray()
        if (text[0] == '.') break

        val st = mutableListOf<Char>()
        var res = "yes"

        for (t in text) {
            if (t == '.') break
            else if (t == '(' || t == '[') st.add(t)
            else if (t == ')' || t == ']') {
                if (st.isEmpty()) res = "no"
                else {
                    val top = st.removeLast()
                    if ((top == '[' && t == ')') || (top == '(' && t == ']')) {
                        res = "no"
                    }
                }
            }
            if (res == "no") break
        }
        res = if (st.isEmpty()) res else "no"
        println(res)
    }
}

File: by-python/boj/4949-균형잡힌 세상.py
if __name__ == '__main__':
    while True:
        text = input()
        if text == '.': break

        st = []
        res = 'yes'
        for t in text:
            if t == '.':
                break
            elif t in ['(', '[']:
                st.append(t)
            elif t in [')', ']']:
                if len(st) == 0:
                    res = 'no'
                else:
                    top = st.pop()
                    if (top == '[' and t == ')') or \
                            (top == '(' and t == ']'):
                        res = 'no'
            if res == 'no': break
        print(res if len(st) == 0 else 'no')
